File: typescript/src/services/readjust/validations/periodicity-between-readjusts.validation.ts
import { differenceInCalendarMonths } from "date-fns";

import { AdjustValidation } from "../../../interfaces/adjust-validation";
import { Employee } from "../../../models/employee";

export class PeriodicityBetweenReadjustsValidation implements AdjustValidation {
  validate(employee: Employee, increase: number) {
    const lastReadjust = employee.lastReadjustDate;
    const actualDate = new Date();
    const monthsOfTheLastReadjust = differenceInCalendarMonths(
      lastReadjust,
      actualDate
    );

    if (monthsOfTheLastReadjust < 6) {
      throw new Error(
        "Interval between the salary adjusts need to be at least 6 months"
      );
    }
  }
}

File: java/src/main/java/br/com/rh/services/readjust/validations/PeriodicityBetweenReadjustsValidation.java
package br.com.rh.services.readjust.validations;

import br.com.rh.exceptions.SalaryException;
import br.com.rh.interfaces.AdjustValidation;
import br.com.rh.models.Employee;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;

public class PeriodicityBetweenReadjustsValidation implements AdjustValidation {
    public void validate(Employee employee, BigDecimal increase) {
        var lastReadjust = employee.getLastReadjustDate();
        var actualDate = LocalDate.now();
        var monthsOfTheLastReadjust = ChronoUnit.MONTHS.between(lastReadjust, actualDate);

        if (monthsOfTheLastReadjust < 6) {
            throw new SalaryException("Interval between the salary adjusts need to be at least 6 months");
        }
    }
}
